pkg/util/sketch/topk_test.go

```diff
@@ -0,0 +1,725 @@
+
+	topk, err = NewCMSTopkForCardinality(nil, 100, max)
+	assert.NoError(t, err)
+	for i := 0; i < max; i++ {
+		topk.Observe(strconv.Itoa(i))
+	}
+	_, bigEnough = topk.Cardinality()
+	assert.True(t, bigEnough)
+}
+
+func TestTopK_Merge(t *testing.T) {
+	nStreams := 1000
+	k := 100
+	maxPerStream := 1000
+	events := make([]event, 0)
+	max := int64(0)
+
+	for i := 0; i < nStreams-k; i++ {
+		num := int64(maxPerStream)
+		n := rand.Int63n(num) + 1
+		if n > max {
+			max = n
+		}
+		for j := 0; j < int(n); j++ {
+			events = append(events, event{name: strconv.Itoa(i), count: 1})
+		}
+	}
+	// then another set of things more than the max of the previous entries
+	for i := nStreams - k; i < nStreams; i++ {
+		n := int64(rand.Int63n(int64(maxPerStream)) + 1 + max)
+		for j := 0; j < int(n); j++ {
+			events = append(events, event{name: strconv.Itoa(i), count: 1})
+		}
+	}
+
+	rand.Seed(time.Now().UnixNano())
+	rand.Shuffle(len(events), func(i, j int) { events[i], events[j] = events[j], events[i] })
+
+	topk1, err := NewCMSTopkForCardinality(nil, k, nStreams)
+	topk2, err := NewCMSTopkForCardinality(nil, k, nStreams)
+	assert.NoError(t, err, "error creating topk")
+	for i := 0; i < (len(events) / 2); i++ {
+		for j := 0; j < events[i].count; j++ {
+			topk1.Observe(events[i].name)
+		}
+	}
+
+	for i := len(events) / 2; i < len(events); i++ {
+		for j := 0; j < events[i].count; j++ {
+			topk2.Observe(events[i].name)
+		}
+	}
+
+	topk1.Merge(topk2)
+
+	mergedTopk := topk1.Topk()
+	var eventName string
+	mergedMissing := 0
+outer:
+	for i := nStreams - k; i < nStreams; i++ {
+		eventName = strconv.Itoa(i)
+		for j := 0; j < len(mergedTopk); j++ {
+			if mergedTopk[j].Event == eventName {
+				continue outer
+			}
+		}
+
+		mergedMissing++
+	}
+
+	require.LessOrEqualf(t, mergedMissing, 2, "more than acceptable misses: %d > %d", mergedMissing, 2)
+
+	// observe the same events into a single sketch
+	topk3, _ := NewCMSTopkForCardinality(nil, k, nStreams)
+	for _, e := range events {
+		for j := 0; j < e.count; j++ {
+			topk3.Observe(e.name)
+		}
+	}
+
+	singleTopk := topk3.Topk()
+
+	singleMissing := 0
+outer2:
+	for i := nStreams - k; i < nStreams; i++ {
+		eventName = strconv.Itoa(i)
+		for j := 0; j < len(singleTopk); j++ {
+			if singleTopk[j].Event == eventName {
+				continue outer2
+			}
+		}
+
+		singleMissing++
+	}
+
+	require.LessOrEqualf(t, singleMissing, 2, "more than acceptable misses: %d > %d", singleMissing, 2)
+	// this condition is never actually true
+	//require.LessOrEqualf(t, mergedMissing, singleMissing, "merged sketch should be at least as accurate as a single sketch")
+}
+
+// compare the accuracy of cms topk and hk to the real topk
+func TestRealTopK(t *testing.T) {
+	// the HLL cardinality estimate for this page is ~72000
+	link := "https://www.gutenberg.org/cache/epub/100/pg100.txt"
+
+	resp, err := http.Get(link)
+	require.NoError(t, err)
+	scanner := bufio.NewScanner(resp.Body)
+
+	m := make(map[string]uint32)
+	h := MinHeap{}
```

Comment:
I'm not quite sure why we are not using a hash map instead to get the actual count.

Reply:
We are, the map `m` tracks the real counts for things, and the heap `h` tracks the current topk. Same logic as the cms implementation, if we observe an event into the heap and it's new count is greater than the heap min we insert that thing into the heap. This should be faster than doing a sort of a potentially massive cardinality set (think 1 million plus unique streams) at the end of observing all the events.

Reply:
Well, you are sorting after all wir haben been scanned or am I missing something?

Reply:
I'm sorting the list I generate from the heap, since iterating over the heap elements it won't be in `top100-top1` order, the only guarantee is that the smallest count element will be first.